datafusion/src/execution/dataframe_impl.rs

```diff
@@ -62,6 +68,60 @@ impl DataFrameImpl {
     }
 }
 
+#[async_trait]
+impl TableProvider for DataFrameImpl {
```

Comment:
perhaps we can add `TableProvider` as one of the trait bounds for the `Dataframe` trait so the trait object can be used to register tables without casting to a concrete type.

Reply:
Can you give an example of what you mean?

It won't be possible to register `Arc<dyn DataFrame>`s directly even if `DataFrame` has a trait bound of `TableProvider`, because that would require trait upcasting IIUC, which is still an unstable feature.

Reply:
I think @houqp  was suggesting changing

```rust
trait DataFrame {
...
}
```

To something like

```rust
trait DataFrame:: TableProvider {
}
```

And then `impl TableProvider` for `DataFrame` (rather than `DataFrameImpl`) 

If this is interesting, I am happy to file a ticket describing it

Reply:
Interesting, I don't think you can impl a trait for a trait (since the introduction of the `dyn` requirement) but we might be able to do a blanket impl like

```rust
impl<D: DataFrame> TableProvider for D {
    ...
}
```

Along with the trait bound, I think this will work to allow any DataFrame implementation to be used as a table provider.

The main question for me is:  given that the `scan` method returns a physical plan, do I need to add a new DataFrame method to get a physical plan? Or is there another way to get that with existing APIs?

Reply:
I think I may have a prototype, I'll push something up here in a bit.

Reply:
I don't think there's any way to achieve this behavior

```rust
        let df = test_table().await?.select_columns(&["c1", "c12"])?;
        let mut ctx = ExecutionContext::new();

        // register a dataframe as a table
        ctx.register_table("test_table", df)?;
```
without `trait_upcasting` 😞, even with a blanket `impl` plus an additional trait bound. I'll put up a draft PR.